development_bank_wales/pipeline/feature_preparation/data_aggregation.py

```diff
@@ -0,0 +1,137 @@
+"""Aggregate data on hex or local authority level."""
+
+# ---------------------------------------------------------------------------------
+
+import pandas as pd
+
+from asf_core_data.getters.supplementary_data.deprivation import imd_data
+from asf_core_data.getters.supplementary_data.geospatial import coordinates
+from asf_core_data.utils.geospatial import data_agglomeration
+
+from development_bank_wales.pipeline.feature_preparation.legacy import upgrades
+
+# ---------------------------------------------------------------------------------
+
+
+def get_supplementary_data(df, data_path="S3"):
+    """Get coordinates and IMD data. Geographical data is required for aggregation.
+
+    Args:
+        df (pd.DataFrame): Dataframe to which supplementary data is added.
+        data_path (str, optional): Data source: local dir or S3. Defaults to "S3".
+
+    Returns:
+        df_suppl: Dataframe with added supplementary data.
+    """
+
+    # Get coordinates and hex ids: resolution of 6 equals 36km2
+    coordinates_df = coordinates.get_postcode_coordinates(data_path=data_path)
+    df_suppl = pd.merge(df, coordinates_df, on="POSTCODE", how="left")
+    df_suppl = data_agglomeration.add_hex_id(df_suppl, resolution=6)
+
+    # Get Wales IMD data
+    wales_imd = imd_data.get_country_imd_data("Wales", data_path=data_path)[
+        ["Postcode", "IMD Decile"]
+    ]
+
+    df_suppl = imd_data.merge_imd_with_other_set(
+        wales_imd, df_suppl, postcode_label="POSTCODE"
+    )
+
+    return df_suppl
+
+
+def get_mean_per_group(features_df, label_set, agglo_f="hex_id"):
+    """Get mean upgrade probability for each group, e.g. hex id or local authority label.
+    Additionally, compute the IMD mean and number of properties per group.
+
+    Args:
+        features_df (pd.DataFrame): Features including upgrade probabilities, e.g. ROOF_proba.
+        label_set (list): Label list, e.g. upgrade categories ROOF, WALLS or FLOOR.
+        agglo_f (str, optional): Feature by which to group. Defaults to "hex_id".
+
+    Returns:
+        group_probas (pd.DataFrame): upgrade probabilities per group.
+    """
+
+    weight_dict = {
+        "ROOF_UPGRADABILITY": 0.35,
+        "WALLS_UPGRADABILITY": 0.5,
+        "FLOOR_UPGRADABILITY": 0.15,
+    }
+
+    group_probas = (
+        features_df.groupby([agglo_f])[["LMK_KEY"]]
+        .count()
+        .reset_index()
+        .rename(columns={"LMK_KEY": "# Properties"})
+    )
+
+    group_probas["IMD Decile (mean)"] = (
+        features_df.groupby([agglo_f])[["IMD Decile"]]
+        .mean()
+        .reset_index()["IMD Decile"]
+    )
+
+    group_probas["IMD Decile (mean)"] = group_probas["IMD Decile (mean)"].round(0)
+
+    for label in label_set:
+
+        proba_label = "proba {}".format(label)
+        group_probas[proba_label] = (
+            features_df.groupby([agglo_f])[[proba_label]]
+            .mean()
+            .reset_index()[proba_label]
+        )
```

Comment:
Instead of doing the `for` loop we can do the `groupby()` for all labels at once. Something like this:
```suggestion
    rename_labels_dict = ...
    group_probas[rename_labels_dict] = (
        features_df.groupby([agglo_f])[label_set]
        .mean()
        .reset_index()[label_set]
    )
```

Actually you could do the `groupby().mean()` together for `IMD Decile` and all labels in `label_set`.